rust/lance/src/index/vector/ivf/builder.rs

```diff
@@ -15,109 +15,109 @@
-};
+use arrow_array::types::UInt32Type;
+use arrow_array::{cast::AsArray, types::Float32Type, UInt32Array};
 use arrow_schema::{DataType, Field, Schema};
-use futures::{
-    stream::{self, repeat_with},
-    StreamExt, TryStreamExt,
-};
-use lance_arrow::{FixedSizeListArrayExt, RecordBatchExt};
+use futures::{stream::repeat_with, StreamExt, TryStreamExt};
+use lance_arrow::RecordBatchExt;
 use lance_core::io::Writer;
 use lance_index::vector::{
     pq::{transform::PQTransformer, ProductQuantizer},
-    transform::Transformer,
-    PQ_CODE_COLUMN,
+    residual::ResidualTransform,
+    PART_ID_COLUMN, PQ_CODE_COLUMN,
 };
 use lance_linalg::{distance::MetricType, MatrixView};
 use snafu::{location, Location};
 use tracing::instrument;
 
 use crate::dataset::ROW_ID;
-use crate::index::vector::ivf::{io::write_index_partitions, shuffler::ShufflerBuilder, Ivf};
+use crate::index::vector::ivf::{
+    io::write_index_partitions,
+    shuffler::{Shuffler, ShufflerBuilder},
+    Ivf,
+};
 use crate::{io::RecordBatchStream, Error, Result};
 
 use super::RESIDUAL_COLUMN;
 
-/// Filter a batch by a range of partition IDs, specified by `part_range`.
-///
-/// Expect the input batch has schema as
+/// Disk-based shuffle a strema of [RecordBatch] into each IVF partition.
+/// Sub-quantizer will be applied if provided.
 ///
-/// ```json
-/// {
-///     "_rowid": "uint32",
-///     "<column>": "fixed_size_list<float16/32/64>",
-/// }
-/// ```
+/// Parameters
+/// ----------
+///   *data*: input data stream.
+///   *ivf*: IVF model.
 ///
-/// And output batch has schema as
+/// Returns
+/// -------
+///   Shuffler: a shuffler that stored the shuffled data.
 ///
-/// ```json
-/// {
-///     "_rowid": "uint32",
-///     "<column>": "fixed_size_list<float16/32/64>",
-///     "__ivf_part_id": "uint32",
-/// }
-/// ```
-fn filter_batch_by_partition(
-    batch: &RecordBatch,
+/// TODO: move this to `lance-index` crate.
+pub(crate) async fn shuffle_dataset(
+    data: impl RecordBatchStream + Unpin,
     column: &str,
-    ivf: &Ivf,
-    metric_type: MetricType,
-    part_range: Range<u32>,
-) -> Result<Vec<(u32, RecordBatch)>> {
-    let arr = batch
-        .column_by_name(column)
-        .expect("The caller already checked column exist")
-        .as_fixed_size_list();
-    let dim = arr.value_length() as usize;
-
-    let matrix = MatrixView::<Float32Type>::try_from(arr)?;
-    let part_ids = ivf.compute_partitions(&matrix, metric_type);
-    // A map from partition ID and row IDs.
-    let mut parted_map: BTreeMap<u32, Vec<u32>> = BTreeMap::new();
-    part_ids
-        .values()
-        .iter()
-        .enumerate()
-        .for_each(|(idx, &part_id)| {
-            if part_range.contains(&part_id) {
-                parted_map.entry(part_id).or_default().push(idx as u32);
-            }
-        });
+    ivf: Arc<lance_index::vector::ivf::Ivf>,
+    // TODO: make this auto-generated from transformer chain.
```

Comment:
Make what auto-generated? `num_sub_vectors`?

Reply:
The end of the transform chain should produce a schema. Lemme add some comments.

For example,  Using BQ will lead to different schema than PQ.